metagraph/src/graph/representation/base/dbg_wrapper.hpp

```diff
@@ -0,0 +1,325 @@
+#ifndef __DBG_WRAPPER__
+#define __DBG_WRAPPER__
+
+#include "sequence_graph.hpp"
+
+
+namespace mtg {
+namespace graph {
+
+/**
+ * This abstract class stores a graph internally and transfers all calls to it.
+ * It can be used as a parent to other wrappers which add additional functionality.
+ * All children of this class must implement add_sequence, load, and operator==
+ */
+template <class Graph = DeBruijnGraph>
+class DBGWrapper : public DeBruijnGraph {
+  public:
+    template <class InGraph>
+    explicit DBGWrapper(std::shared_ptr<InGraph> graph)
+          : graph_ptr_(std::dynamic_pointer_cast<Graph>(graph)), graph_(graph_ptr_) {
+        assert(graph_);
+        assert(graph_ptr_);
+    }
+
+    template <class InGraph>
+    explicit DBGWrapper(std::shared_ptr<const InGraph> graph)
+          : graph_(std::dynamic_pointer_cast<const Graph>(graph)) {
+        assert(graph_);
+        assert(!graph_ptr_);
+    }
+
+    // aliasing constructor
+    template <class InGraph>
+    explicit DBGWrapper(const InGraph &graph)
+          : graph_(std::shared_ptr<const Graph>{}, dynamic_cast<const Graph*>(&graph)) {
+        assert(graph_);
+        assert(!graph_ptr_);
+    }
+
+    virtual ~DBGWrapper() {}
+
+    /**
+     * Added methods
+     */
+    virtual const Graph& get_graph() const { return *graph_; }
+    virtual std::shared_ptr<const Graph> get_graph_ptr() const { return graph_; }
+    virtual std::shared_ptr<Graph> get_mutable_graph_ptr() const { return graph_ptr_; }
```

Comment:
Any reason to have this mutable version as well? I think this is not needed at the moment and can be removed for simplicity.

add_sequence should just throw errors

Reply:
For wrappers like `DBGSuccinctCached` which don't modify the external view of the graph and only keep an internal cache, allowing for non-const graphs to work has very little extra overhead.

Reply:
Yes, but we aren't using this. So it's still better to remove this.

Reply:
The only place where it's used is in the unit tests for `DBGSuccinctCached` (since I add the cached version as another graph type)

Reply:
Maybe just test it like all other static graphs?

Reply:
We don't really have a framework for testing these other than copying and pasting unit tests, so it's not ideal.

Reply:
That can be saved for a later pull request.

Reply:
but how do we test DBGBitmap, for instance?

Reply:
It's still modifiable in the sense that `load` is implemented. So when adding unit tests for `DBGSuccinctCached`, I had to implement serialize and load. At that point, I figured why not also implement add_sequence since it's trivial

Reply:
So the graph pointer had to be mutable to implement load.

Reply:
I see, but we can add many things, though each of them makes the code more complex and harder to extend. So if those features are not used, it's better not to add them, even if it looks natural.

Do we actually call load on these wrappers? I thought we always load the graph and only then wrap it into something else.

Reply:
load and serialize is only used in the unit tests. One could add another set of graphs called StaticGraphs to test_dbg_helpers so that the load and serialize tests are only run for non-const graph types.

Reply:
we didn't have load/serialize in CanonicalDBG, but you still were able to test it. Why don't we keep things as they were?

Reply:
Anyway, for this pull request we don't need a non-const pointer, and for #334 one can do the unit tests without needing to add a non-const pointer, so I just removed this feature...